### PR TITLE
Move key equality functor and empty value sentinel to OA impl class

### PR DESCRIPTION
**Pull request by sleeepyjack**
This PR moves the logic for handling the key equality predicate and the `empty_value_sentinel` (for `static_map`)  to the OA impl class.

### CODE REVIEW

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 1]
```diff
@@ -113,20 +120,76 @@ class open_addressing_ref_impl {
...
    * @brief Constructs open_addressing_ref_impl.
    *
+   * @param empty_key_sentinel The reserved key value for empty slots
    * @param empty_slot_sentinel Sentinel indicating an empty slot
+   * @param predicate Key equality binary callable
    * @param probing_scheme Probing scheme
    * @param storage_ref Non-owning ref of slot storage
    */
   __host__ __device__ explicit constexpr open_addressing_ref_impl(
+    cuco::empty_key<key_type> empty_key_sentinel,
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: This is another cleanup I wanted to bring up: with ref having `has_payload`, the key sentinel is no longer necessary since the info is already contained in `empty_slot_sentinel_`

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 2]
```diff
@@ -679,53 +691,35 @@ class open_addressing_ref_impl {
...
-  /**
-   * @brief Gets the non-owning storage ref.
+   * @brief Extracts the key from a given value type.
    *
-   * @return The non-owning storage ref of the container
+   * @return The key
    */
-  [[nodiscard]] __device__ constexpr storage_ref_type storage_ref() const noexcept
+  template <typename Value>
+  [[nodiscard]] __host__ __device__ constexpr auto extract_key(Value const& value) const noexcept
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
  [[nodiscard]] __device__ constexpr auto extract_key(Value const& value) const noexcept
```
nit: probably device only
Also, would returning by ref make a difference?

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 3]
```diff
@@ -850,33 +840,31 @@ class open_addressing_ref_impl {
...
-      return back_to_back_cas(slot, value, predicate);
+      return back_to_back_cas(slot, value);
 #endif
     }
   }
 
-  value_type empty_slot_sentinel_;      ///< Sentinel value indicating an empty slot
-  probing_scheme_type probing_scheme_;  ///< Probing scheme
-  storage_ref_type storage_ref_;        ///< Slot storage ref
+  value_type empty_slot_sentinel_;  ///< Sentinel value indicating an empty slot
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]:
```suggestion
```
This can be removed also since the sentinel is wrapped in `predicate_` already

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 1 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: To clarify, this can be deferred to another PR if you want.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 4]
```diff
@@ -113,20 +120,76 @@ class open_addressing_ref_impl {
...
+  {
+    return empty_slot_sentinel_;
+  }
+
+  /**
+   * @brief Returns the function that compares keys for equality.
+   *
+   * @return The key equality predicate
+   */
+  [[nodiscard]] __host__ __device__ constexpr detail::equal_wrapper<key_type, key_equal> key_eq()
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
  [[nodiscard]] __host__ __device__ constexpr detail::equal_wrapper<key_type, key_equal> const& key_eq()
```
Similar to `probing_scheme`, probably returns by `const&` instead of by value.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 3 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]: I think the predicate only holds the key sentinel, not the slot sentinel, which can be a pair of `{key_sentinel, value_sentinel}`.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: It's a small change so it fits in this PR. I'm working on it. BTW good idea! Much cleaner.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 2 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: Benchmarking...

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 4 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: Should we make `empty_key_sentinel()` and `empty_slot_sentinel()` return `const&`, too? If we add support for larger types, this might become beneficial. I'll benchmark this one too.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]: Yeah, that's fine for this PR. Shortly, we will add the erased sentinel into this equal wrapper as well thus we are duplicating more and more vars.

Can you please put a TODO reminding potential cleanups for sentinels in both wrapper and container ref?

Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]: Yes, will do. Good point about the `erased_key_sentinel`. We somehow need to rethink this design once we add support for `erase`, since now there are two possible slot sentinels.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 4 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: Equal wrapper and probing scheme (probably sentinel slot also) would be always returned by ref since it's used internally only and we know copying is not necessary.

As for key sentinel (or haser or key_eq), depending on the use case, if it's a public API in data structures, returning by value aligns with std behavior and the performance wouldn't matter in that case. If it's in the implementation details (like here in the OA ref base class), maybe returning by ref is preferable?

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]: > since now there are two possible slot sentinels.

That's what I'm struggling with for erase

File: include/cuco/detail/static_map/static_map_ref.inl [hunk 5]
```diff
@@ -355,7 +268,7 @@ class operator_impl<
 
       auto const [state, intra_window_index] = [&]() {
         for (auto i = 0; i < window_size; ++i) {
-          switch (ref_.predicate_(window_slots[i], key)) {
+          switch (ref_.impl_.key_eq()(window_slots[i].first, key)) {
```
Comment (PointKernel) [inline include/cuco/detail/static_map/static_map_ref.inl]: That's what I'm worried about, we need the `key_eq` functor just for comparison and returning it by value meaning we copy sentinels every time we perform an equality check. That's really expensive.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 4 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: It's public API for `static_map` and `static_set`, so those function need to comply with the STL. The OA impl could return a reference, since it is only used internally.

Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: The benchmark results are a bit inconclusive:
The only visible effect is on the `insert` side of things.
copy vs. ref as proposed:
# static_set_insert_unique_occupancy

## [0] NVIDIA RTX A6000

|  Key  |  Distribution  |  Occupancy  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |        Diff |   %Diff |  Status  |
|-------|----------------|-------------|------------|-------------|------------|-------------|-------------|---------|----------|
|  I32  |     UNIQUE     |     0.1     |  43.455 ms |       2.83% |  42.950 ms |       3.65% | -504.501 us |  -1.16% |   PASS   |
|  I32  |     UNIQUE     |     0.2     |  43.326 ms |       2.19% |  43.237 ms |       2.76% |  -89.373 us |  -0.21% |   PASS   |
|  I32  |     UNIQUE     |     0.3     |  42.708 ms |       3.17% |  43.196 ms |       2.73% |  488.176 us |   1.14% |   PASS   |
|  I32  |     UNIQUE     |     0.4     |  43.404 ms |       2.42% |  43.308 ms |       2.60% |  -96.021 us |  -0.22% |   PASS   |
|  I32  |     UNIQUE     |     0.5     |  42.228 ms |       1.75% |  42.870 ms |       2.25% |  641.808 us |   1.52% |   PASS   |
|  I32  |     UNIQUE     |     0.6     |  42.242 ms |       1.31% |  42.090 ms |       1.29% | -152.005 us |  -0.36% |   PASS   |
|  I32  |     UNIQUE     |     0.7     |  42.135 ms |       0.73% |  41.840 ms |       1.36% | -294.600 us |  -0.70% |   PASS   |
|  I32  |     UNIQUE     |     0.8     |  43.044 ms |       1.08% |  43.218 ms |       0.05% |  174.129 us |   0.40% |   FAIL   |
|  I32  |     UNIQUE     |     0.9     |  45.407 ms |       0.08% |  45.033 ms |       0.99% | -373.296 us |  -0.82% |   FAIL   |
|  I64  |     UNIQUE     |     0.1     |  45.105 ms |       3.22% |  44.919 ms |       2.98% | -186.440 us |  -0.41% |   PASS   |
|  I64  |     UNIQUE     |     0.2     |  44.373 ms |       2.86% |  43.840 ms |       3.39% | -533.816 us |  -1.20% |   PASS   |
|  I64  |     UNIQUE     |     0.3     |  44.851 ms |       1.87% |  44.328 ms |       2.68% | -523.048 us |  -1.17% |   PASS   |
|  I64  |     UNIQUE     |     0.4     |  44.468 ms |       2.81% |  44.831 ms |       3.18% |  362.698 us |   0.82% |   PASS   |
|  I64  |     UNIQUE     |     0.5     |  44.881 ms |       2.66% |  44.896 ms |       3.74% |   15.506 us |   0.03% |   PASS   |
|  I64  |     UNIQUE     |     0.6     |  45.511 ms |       1.78% |  45.505 ms |       1.97% |   -5.970 us |  -0.01% |   PASS   |
|  I64  |     UNIQUE     |     0.7     |  46.298 ms |       2.09% |  46.335 ms |       1.89% |   36.448 us |   0.08% |   PASS   |
|  I64  |     UNIQUE     |     0.8     |  46.323 ms |       1.12% |  46.208 ms |       1.26% | -114.997 us |  -0.25% |   PASS   |
|  I64  |     UNIQUE     |     0.9     |  48.824 ms |       0.14% |  48.719 ms |       0.94% | -105.239 us |  -0.22% |   FAIL   |

Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: If we addidionally make the `empty_*_sentinel()` functions return a ref we get:
copy vs. everything ref:
# static_set_insert_unique_occupancy

## [0] NVIDIA RTX A6000

|  Key  |  Distribution  |  Occupancy  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |        Diff |   %Diff |  Status  |
|-------|----------------|-------------|------------|-------------|------------|-------------|-------------|---------|----------|
|  I32  |     UNIQUE     |     0.1     |  43.455 ms |       2.83% |  43.161 ms |       3.15% | -294.272 us |  -0.68% |   PASS   |
|  I32  |     UNIQUE     |     0.2     |  43.326 ms |       2.19% |  43.069 ms |       3.12% | -257.351 us |  -0.59% |   PASS   |
|  I32  |     UNIQUE     |     0.3     |  42.708 ms |       3.17% |  42.461 ms |       3.58% | -247.056 us |  -0.58% |   PASS   |
|  I32  |     UNIQUE     |     0.4     |  43.404 ms |       2.42% |  43.318 ms |       2.82% |  -86.169 us |  -0.20% |   PASS   |
|  I32  |     UNIQUE     |     0.5     |  42.228 ms |       1.75% |  42.588 ms |       2.92% |  359.549 us |   0.85% |   PASS   |
|  I32  |     UNIQUE     |     0.6     |  42.242 ms |       1.31% |  43.028 ms |       1.01% |  786.802 us |   1.86% |   FAIL   |
|  I32  |     UNIQUE     |     0.7     |  42.135 ms |       0.73% |  42.703 ms |       0.04% |  567.946 us |   1.35% |   FAIL   |
|  I32  |     UNIQUE     |     0.8     |  43.044 ms |       1.08% |  43.570 ms |       1.12% |  526.099 us |   1.22% |   FAIL   |
|  I32  |     UNIQUE     |     0.9     |  45.407 ms |       0.08% |  45.788 ms |       0.06% |  381.789 us |   0.84% |   FAIL   |
|  I64  |     UNIQUE     |     0.1     |  45.105 ms |       3.22% |  45.367 ms |       2.33% |  261.648 us |   0.58% |   PASS   |
|  I64  |     UNIQUE     |     0.2     |  44.373 ms |       2.86% |  44.203 ms |       3.42% | -170.016 us |  -0.38% |   PASS   |
|  I64  |     UNIQUE     |     0.3     |  44.851 ms |       1.87% |  44.338 ms |       2.47% | -513.147 us |  -1.14% |   PASS   |
|  I64  |     UNIQUE     |     0.4     |  44.468 ms |       2.81% |  45.275 ms |       2.21% |  807.341 us |   1.82% |   PASS   |
|  I64  |     UNIQUE     |     0.5     |  44.881 ms |       2.66% |  45.939 ms |       2.23% |    1.058 ms |   2.36% |   FAIL   |
|  I64  |     UNIQUE     |     0.6     |  45.511 ms |       1.78% |  45.582 ms |       1.56% |   70.513 us |   0.15% |   PASS   |
|  I64  |     UNIQUE     |     0.7     |  46.298 ms |       2.09% |  46.162 ms |       3.14% | -136.134 us |  -0.29% |   PASS   |
|  I64  |     UNIQUE     |     0.8     |  46.323 ms |       1.12% |  46.207 ms |       1.47% | -115.657 us |  -0.25% |   PASS   |
|  I64  |     UNIQUE     |     0.9     |  48.824 ms |       0.14% |  48.804 ms |       0.83% |  -20.512 us |  -0.04% |   PASS   |

**Comment by sleeepyjack** [comment]
One thing I'm not 100% sure I did right and thus like to have a second pair of eyes to verify: Did I map the comparison operators correctly, i.e., only use the three-way comparison operator when necessary?

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: Addressed in [dc71ba0](https://github.com/NVIDIA/cuCollections/pull/379/commits/dc71ba08aa81c942c0537795c228bb937c16b9ce)

File: include/cuco/detail/static_map/static_map_ref.inl (hunk 5 above)
Comment (sleeepyjack) [reply include/cuco/detail/static_map/static_map_ref.inl]: Fixed it in [36d1d5d](https://github.com/NVIDIA/cuCollections/pull/379/commits/36d1d5d2b51b40fd2b36efe6e451773c4c84a397)

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 6]
```diff
@@ -100,8 +105,9 @@ class open_addressing_ref_impl {
   using value_type          = typename storage_ref_type::value_type;   ///< Storage element type
   using extent_type         = typename storage_ref_type::extent_type;  ///< Extent type
   using size_type           = typename storage_ref_type::size_type;    ///< Probing scheme size type
-  using iterator            = typename storage_ref_type::iterator;     ///< Slot iterator type
-  using const_iterator = typename storage_ref_type::const_iterator;    ///< Const slot iterator type
+  using key_equal           = KeyEqual;  ///< Type of key equality binary callable
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:108]: need to update the `key_equal` aliases in set_ref and map_ref as well, something like
```cpp
using key_equal = typename impl_type::key_equal;
```

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]:
```suggestion
  // TODO: Clean up the sentinel handling since it's duplicated in ref and equal wrapper
  value_type empty_slot_sentinel_;  ///< Sentinel value indicating an empty slot
```

File: include/cuco/detail/static_map/static_map_ref.inl [hunk 7]
```diff
@@ -309,7 +221,7 @@ class operator_impl<
       auto const window_slots = storage_ref[*probing_iter];
 
       for (auto& slot_content : window_slots) {
-        auto const eq_res = ref_.predicate_(slot_content, key);
+        auto const eq_res = ref_.impl_.key_eq()(slot_content.first, key);
```
Comment (PointKernel) [inline include/cuco/detail/static_map/static_map_ref.inl]: This is slightly confusing since the `key_eq` (following the std naming convention) public API returns the user-provided equality functor while here it's actually our equal wrapper. My personal preference would be `predicate()` but not a hard requirement though.

File: include/cuco/detail/static_map/static_map_ref.inl [hunk 8]
```diff
@@ -44,9 +44,11 @@ __host__ __device__ constexpr static_map_ref<
...
                                 ProbingScheme const& probing_scheme,
                                 StorageRef storage_ref) noexcept
-  : impl_{cuco::pair{empty_key_sentinel, empty_value_sentinel}, probing_scheme, storage_ref},
-    empty_value_sentinel_{empty_value_sentinel},
-    predicate_{empty_key_sentinel, predicate}
+  : impl_{cuco::pair{empty_key_sentinel, empty_value_sentinel},
+          predicate,
+          probing_scheme,
+          storage_ref},
+    empty_value_sentinel_{empty_value_sentinel}
```
Comment (PointKernel) [inline include/cuco/detail/static_map/static_map_ref.inl]: could add `extract_payload` utility into the OA ref base class so `empty_value_sentinel_` doesn't have to be a member of map_ref. Could be a separate PR.

**Review by PointKernel — COMMENTED** [review]
Really nice cleanups! Love the changes. 

`key_equal` alias definition should be unified and other suggestions are not mandatory.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 6 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:108]: I played around with this a bit and the problem is a bit more complex, since we need to deduce some of the alias types from the ctor using CTAD. It somehow works for now, but we probably need to revisit this in a separate PR. If I apply your suggestion somehow everything goes up in flames -.-

**Comment by sleeepyjack** [comment]
@PointKernel all change requests have been addressed

File: include/cuco/detail/static_map/static_map_ref.inl (hunk 8 above)
Comment (sleeepyjack) [reply include/cuco/detail/static_map/static_map_ref.inl]: Addressed in [5a99a56](https://github.com/NVIDIA/cuCollections/pull/379/commits/5a99a56a81d6866b0ae1ecf3a669d64b2916f723)

File: include/cuco/detail/static_map/static_map_ref.inl (hunk 7 above)
Comment (sleeepyjack) [reply include/cuco/detail/static_map/static_map_ref.inl]: Addressed in [35047c4](https://github.com/NVIDIA/cuCollections/pull/379/commits/35047c482139bef43a340090aba05ba6d07d7b08)

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 3 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:896]: Addressed in [c390393](https://github.com/NVIDIA/cuCollections/pull/379/commits/c39039304418523d1073c2a67ff3911e49839824)

**Review by PointKernel — APPROVED** [review]
Looks great! Thanks!

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 6 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:108]: I think I see what you mean, I've encountered similar issues back then.
